src/pudl/output/ferc1.py

```diff
@@ -1063,6 +1077,88 @@ def calculations(self):
+          the explosion should be turned into leaves -- i.e. they should be replaced
+          with a single calculation component filled with NA values.
+        - Every remaining calculation component must also appear as a parent (if it
+          is a leaf, then it will have a single null calculation component)
+        - There should be no records where only one of table_name or xbrl_factoid
+          are null. They should either both or neither be null.
+        - table_name_parent and xbrl_factoid_parent should be non-null.
+        """
+        calc_cols = list(NodeId._fields)
+        parent_cols = [col + "_parent" for col in calc_cols]
+
+        # Keep only records where parent is within exploded tables:
+        calc_explode = self.calculation_components_xbrl_ferc1[
+            self.calculation_components_xbrl_ferc1.table_name_parent.isin(
+                self.table_names
+            )
+        ].copy()
+        # Groupby parent factoids
+        gb = calc_explode.groupby(
+            ["table_name_parent", "xbrl_factoid_parent"], as_index=False
+        )
+        # Identify groups where all calculation components are within the explosion
+        # We need to include table_name of pd.NA to retain non-calculated values.
+        calc_explode["is_within_explosion"] = gb.table_name.transform(
+            lambda x: x.isin(self.table_names + [pd.NA]).all()
+        )
+        # Convert any calculation componets from outside the explosion into reported
+        # values / leaf nodes:
+        calc_explode.loc[
+            ~calc_explode.is_within_explosion, list(NodeId._fields) + ["weight"]
+        ] = pd.NA
+        calc_explode = calc_explode.drop(columns=["is_within_explosion"]).reset_index(
+            drop=True
+        )
+
+        ##############################################################################
+        # Everything below here is error checking / debugging
+
+        # There should be no cases where only one of table_name or xbrl_factoid is NA:
+        partially_null = calc_explode[
+            calc_explode.table_name.isna() ^ calc_explode.xbrl_factoid.isna()
+        ]
+        if not partially_null.empty:
+            logger.error(
+                "Found unexpectedly null calculation components:\n"
+                f"{partially_null[['table_name', 'xbrl_factoid']]}"
+            )
+
+        # Using list of NodeIds rather than index here due to issues with NA types and
+        # it also avoids needing to rename columns or index level names.
+        # An index of all non-null child nodes / calculation components:
+        calc_idx = [
+            NodeId(*x)
+            for x in calc_explode[calc_cols].dropna(how="all").itertuples(index=False)
+        ]
+        # An index of all parent nodes:
+        parent_idx = [
+            NodeId(*x)
+            for x in calc_explode[parent_cols].dropna(how="all").itertuples(index=False)
+        ]
```

Comment:
you can't drop all nulls because it is very expected that the dimension columns will have nulls in them

Reply:
higher level q: is there a benefit to converting these into nodes as opposed to just setting/using the index of the df?

Reply:
oh sorry i get the "all" now! i was assuming this was how="any"

Reply:
Doesn't `how="all"` mean it'll only drop records where *all* the values are null?

Reply:
I mentioned this in the comment, but there are 2 reasons for using the node list here:
* It has `pd.NA` rather than `np.nan` which is required to use it to select values (this is a bit of a hack and I'm sure I can make the index work this way too. It's an index data type thing).
* The list of nodes doesn't know what set of columns it applies to, so the same list can be used to select either parent or calculation component records, without needing to rename anything.

Reply:
i did this bc simpler/seems easier to grab the outpts but i don't doubt the nulls are a problem
```
calc_idx = calc_explode.dropna(subset=["table_name", "xbrl_factoid"]).set_index(calc_cols).index
# An index of all parent nodes:
parent_idx = calc_explode.dropna(subset=["table_name_parent", "xbrl_factoid_parent"]).set_index(parent_cols).index

# Nodes that *only* appear as children / calculation components:
calc_only = calc_idx.difference(parent_idx)
```